docs/vocabulary.md

```diff
@@ -388,6 +389,230 @@ pipeline in corresponding documentation.
 
+### Pipeline Step Types
+
+The table below is an attempt to create a mapping of common names for the various types of Pipeline Steps (the unit of work two degrees smaller than a Pipeline). (The effort of proposing a common vocabulary across CI/CD Tools and Technologies is not yet performed, but we'll use the most common term from the table above as the working term.)
+
+While CI/CD tools and technologies generally give developers broad leeway in naming and implementing their Pipeline Steps, agreeing upon some common terms will help develop more pluggable pipelines. This list is not meant to be exhaustive.
+
+Each type of Pipeline Step will generally have several implementations that are tool specific. For example, Source will have implementations for various Source Code Management (SCM) tools. Publish will have implementations for each type of repository, and so on.
+
+#### Setup
+* Semantics: Provision build resources, set up the build workspace.
+* Aliases: Initialize, Start, Prepare, Workspace
+* Inputs: Build Request Parameters, Build Container Image Name and Version
+* Outputs: Secrets, Build Environment, Build Workspace
+* Other Results and Side Effects: N/A
+
+#### Source
+* Semantics: Download, retrieve or copy software, images, and documentation into the build workspace. Fetch configuration data.
+* Aliases: Clone, Fetch
+* Inputs: Source Code Reference (Repository, Branch, Commit)
+* Outputs: Software Source, Binary Source
+* Other Results and Side Effects: N/A
+
+#### Tag
+* Semantics: Annotate source code, images, etc. with information, such as the version number and a description.
+* Aliases: Annotate, Version
+* Inputs: Source Code Reference (Repository, Branch, Commit), Software Source, Binary Source, Generated Software, Generated Binaries
+* Outputs: Tag/annotation metadata
+* Other Results and Side Effects: Tags/annotations have been added to the software.
+
+#### Secret Detection
+* Semantics: Detect secrets in the source code, other software, or documentation. Examples include passwords, SSH keys, API keys, and so on.
+* Aliases: N/A
+* Inputs: Source Code Reference (Repository, Branch, Commit), Software Source, Binary Source
+* Outputs: Secret Detection Report
+* Other Results and Side Effects: Revoked Secrets
+
+#### Build
+* Semantics: Assemble and/or compile software and documentation into an executable and usable format.
+* Aliases: Compile, Install, Assemble, Generate
+* Inputs: Software Source, Binary Source
+* Outputs: Generated Software, Generated Binaries
+* Other Results and Side Effects: N/A
+
+#### Discovery
+* Semantics: Perform deep discovery to identify all dependencies (including transitive dependencies) in the software and documentation as packaged.
+* Aliases: Dependencies
+* Inputs: Software Source, Binary Source, Generated Software, Generated Binaries
+* Outputs: Dependency List/Graph
+* Other Results and Side Effects: N/A
+
+#### Remediate
```

Comment:
I am quite new to the area of remediation, but I believe I have now heard it used for two related-but-different concepts.
One, as described in this document and in some static code analysis tools, is fixing issues is source code and dependencies before they become a problem.
Another, as used in continuous deployment and cyber security, is taking actions for getting already-running systems to keep running.

Reply:
You're right.  I'm so accustomed to immutable infrastructure (or infrastructure as code, where I don't care if it's a new or old system) that I usually don't think about pipelines patching running systems.

Do you think that would be a different type of step, or should we add that as another example of a Remediate step? For now, I'll add it as an example of the same type of step.

Reply:
For me it is fine to add it as an example of the same type of step, but I should state that I'm really not knowledgeable in the subject.